[CORE-7058]: `storage`: fix race condition in `segment::release_appender_in_background()`

WillemKauf opened the pull request:
This function presented a race between `segment::close()` and a segment roll.

Consider the following sequence of events:

1. `_gate.close()` called in `segment::close()`
2. `auto a = std::exchange(_appender, nullptr)` called in `segment::release_appender_in_background()`
3. `ssx::spawn_with_gate()` called in `segment::release_appender_in_background()`
4. `return ignore_shutdown_exceptions()` in `ssx::spawn_with_gate()`
5. rest of `release_appender_in_background()` is ignored
6. `a` goes out of scope in `release_appender_in_background()` without `close()`ing the `appender`
7. one sad panda:

> segment_appender.cc:78) '_bytes_flush_pending == 0 && _closed' Must flush & close before deleting {no_of_chunks:64, closed:0, fallocation_offset:33554432, stable_offset:5895, flushed_offset:5895, committed_offset:5895, inflight:0, dispatched:0, merged:0, bytes_flush_pending:0}"

Add an explicit check to `_gate.is_closed()` in `release_appender_in_background()` to defer the closing of the appender to `segment::close()` and avoid the potential race condition here.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v24.3.x
- [X] v24.2.x
- [ ] v24.1.x

## Release Notes

### Bug Fixes

* Fixes a bug in which a segment  being rolled and closed could race, leading to a triggered `vassert`.

WillemKauf committed:
`storage`: move `storage_e2e_fixture` to its own header file

WillemKauf committed:
`storage`: add `test_concurrent_segment_roll_and_close`

To test race conditions between `segment::close()` and a segment roll,
particularly one which goes through `release_appender_in_background()`.

andrwng commented on src/v/storage/segment.cc:
```diff
@@ -297,37 +297,29 @@ ss::future<> segment::release_appender(readers_cache* readers_cache) {
-                    });
-              });
-      });
+    ssx::spawn_with_gate(_gate, [this, readers_cache]() mutable {
```
Does this mean we lose the guarantee upon returning in the general case that _appender is null after calling? If the result of calling this when the gate is closed is that it doesn't close _appender anyway, I wonder if we should just keep the old impl, but explicitly check the gate before doing anything

andrwng replied:
>Does this mean we lose the guarantee upon returning in the general case that _appender is null after calling

FWIW I'm not sure if this property is actually important, but it's a notable change in behavior that seems like it _could_ be relied on elsewhere

WillemKauf replied:
>Does this mean we lose the guarantee upon returning in the general case that _appender is null after calling?

Sort of. If you consider the race laid out in the cover letter, this edge case would previously manifest as a triggered `vassert()` immediately. 

In all other cases, `_appender` will be `nullptr` and closed correctly. It is lucky that the general case has never hit this.

If we want to enforce that `_appender` must ` == nullptr` post `release_appender()`, which I think is a very fair contract, we could always move the gate holder call to within `release_appender()` instead? Or maybe the `gate.close()` call till after the write lock is obtained in `segment::close()`?

>If the result of calling this when the gate is closed is that it doesn't close _appender anyway, I wonder if we should just keep the old impl, but explicitly check the gate before doing anything

I do not see how the outcome is different if we early return upon encountering a closed gate, versus silently catch an exception by entering a closed gate with `ssx::spawn_with_gate()`?

andrwng replied:
I mean that if the gate is open, aren't we losing the guarantee that the appender exchange happens during this call? Agreed that we don't have such a guarantee in the general case, but I just want to make sure we're not regressing the common case

WillemKauf replied:
Ah, I see now. I don't think we depend on this behavior currently but I do think it's still a valid contract to uphold.

Another possible change we could make if we didn't want to lose this guarantee is to flip the order of `co_await _gate.close()` and `co_await write_lock()` in `segment::close()`, though that change in behavior could introduce other side effects as well.

What do you think?

andrwng replied:
An early return seems less likely to introduce a bug than ordering changes that involve a lock. I'd prefer the early return because it's more obvious to me it's at least not a regression of the common case

WillemKauf replied:
Thanks for the feedback, I agree.

Changed this to a `_gate.is_closed()` check and an early return.

andrwng replied:
Thanks!

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--01ba021e-7e0f-49b1-b6c8-949476a1e3b8-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59427#01939f0d-e5b3-4448-b9d1-51dae0694acb:
```
"rptest.tests.cluster_config_test.ClusterConfigLegacyDefaultTest.test_legacy_default_explicit_before_upgrade.wipe_cache=False"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59427-->
<!--086ecb8a-a3ca-488f-a04c-e59699d52d42-->
<!--title-->
## Retry command for [Build#59427](https://buildkite.com/redpanda/redpanda/builds/59427)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_config_test.py::ClusterConfigLegacyDefaultTest.test_legacy_default_explicit_before_upgrade@{"wipe_cache":false}
```
<!--/body-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--94b19833-2f25-488b-bad8-609e47b50f55-->
<!--RETRIED FAILURES-->
<!--8330dc92-7413-45ec-8af6-204a6a8e3b72-->
<!--RETRIED FAILURES-->
<!--105b7c7b-0e1c-45ec-b89b-8e50281adead-->
<!--RETRIED FAILURES-->
<!--e1a1fc1e-65f8-402a-8756-987603df3126-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59427#01939fe0-ce7f-4dc5-b353-d352e6575a7a
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59504#0193ad96-74d4-43d4-88ed-6cb4b8f62167
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59622#0193b857-9939-48e4-82be-17d77219ef01
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59622#0193b857-9936-4364-9140-dafe30d38566
<!--/body-->

WillemKauf committed:
`storage`: fix race condition in `segment::release_appender_in_background()`

This function caused race conditions between `segment::close()` and
a segment roll.

Consider the following sequence of events:

1. `_gate.close()` called in `segment::close()`
2. `auto a = std::exchange(_appender, nullptr)` called
    in `segment::release_appender_in_background()`
3. `ssx::spawn_with_gate()` called in `segment::release_appender_in_background()`
4. `return ignore_shutdown_exceptions()` in `ssx::spawn_with_gate()`
5. rest of `release_appender_in_background()` is ignored
6. `a` goes out of scope in `release_appender_in_background()` without
   `close()`ing the `appender`
7. one sad panda

Add an explicit check to `_gate.check()` in `release_appender_in_background()`
to throw an exception in case the gate is closed, and defer the closing of the
appender to `segment::close()` in order to avoid the potential race condition here.

WillemKauf committed:
`storage`: use `model::next_offset()` in `segment::force_roll()`

If the dirty offset is default initialized to `int64_t::min()`
and not updated before a segment is force rolled, an assert will fail
in `disk_log_impl::new_segment()` for the offset being < 0.

Use `model::next_offset()` instead of simply adding `1` to the dirty offset
to avoid this case.

WillemKauf commented:
Force push to:
* Add `_gate.is_closed()` early return path in `segment::release_appender_in_background()`
* Modify `disk_log_impl::force_roll()` to use `model::next_offset()` instead of a possibly error-prone addition operation
* Update `storage_e2e_fixture::remove_segment_permanently()` to also erase `segment` from `log->segments()` to avoid a double `close()`

dotnwat reviewed (commented):
Can you add to the description which fibers are performing which actions? for example

```
F1: _gate.close in segment::close
F2: auto a = std::exchange in segment::release_appender_in_background()
F1: ...
```

like one lingering question is that if the interleaving exists, does it have a race or should it exist at all? for instance, if the segment is being closed, where is the concurrency originating from that allow it to have other fibers operating on it.

dotnwat reviewed (commented):
The race seems plausible, but there doesn't seem to be enough information in the cover letter to retrace the path.

For example, assume that a fiber reached `release_appender_in_background(readers_cache);` via `disk_log_impl::apply_segment_ms` which invokes `last->release_appender`. This fiber would be holding `_compaction_housekeeping_gate`.

So where did `segment::close()` originate from. One option would be `disk_log_impl::close`. But the first thing it does is closes the `_compaction_housekeeping_gate`.

But this interleaving I don't think is possible. Either closing the gate blocks because the release is in progress, or the releasing fiber isn't allowed to enter the gate. Does that sound right?

In order to fully evaluate the fix it's useful to be able to understand how the race occurred.

Can you provide that information in the cover letter?

WillemKauf commented:
>So where did `segment::close()` originate from

From the logs we [have here](https://redpandadata.atlassian.net/browse/CORE-7058?focusedCommentId=49166), it originates from `disk_log_impl::remove_prefix_full_segments()`.

```
TRACE 2024-12-06 16:20:26,087 [shard 0:main] storage - disk_log_impl.cc:1667 - creating log appender for: {redpanda/controller/0}
INFO  2024-12-06 16:20:26,087 [shard 0:main] storage - disk_log_impl.cc:2546 - Removing "/var/lib/redpanda/data/redpanda/controller/0_0/0-1-v1.log" (remove_prefix_full_segments)
```

Will add more detail to the cover letter after I finish sketching the sequence of events out further.

WillemKauf commented:
Huh. Maybe this PR https://github.com/redpanda-data/redpanda/pull/24306 fixed its race condition but introduced one here.

WillemKauf commented:
1. F1: `prefix truncate {redpanda/controller/0}` in `disk_log_impl::do_truncate_prefix()`
2. F2: `creating log appender for {redpanda/controller/0}` in `disk_log_impl::make_appender()`
3. F1: `ptr->write_lock()` in `disk_log_impl::remove_prefix_full_segments()` finally returns, we create a future for the result of `disk_log_impl::remove_segment_permanently()`, but we also drop our `write_lock` before returning the future/waiting for it to finish.
4. F2: `disk_log_appender::initialize()` is then able to obtain a `ptr->read_lock()` while `write_lock` is dropped
5. F1: `Removing "/var/lib/redpanda/data/redpanda/controller/0_0/0-1-v1.log" (remove_prefix_full_segments)` in `disk_log_impl::remove_segment_permanently()`
6. F1: Because of `read_lock` obtained in `appender`, we see: `Segment has outstanding locks. Might take a while to close:/var/lib/redpanda/data/redpanda/controller/0_0/0-1-v1.log` in `disk_log_impl::remove_segment_permanently()`
7. F1: `closing segment: /var/lib/redpanda/data/redpanda/controller/0_0/0-1-v1.log` in `segment::close()`
8. F1: in `segment::close()`, we `co_await _gate.close()`
9. F3: From an unknown origin, we roll the segment and call `segment::release_appender()`
10. F3: We reach `segment::release_appender_in_background()` because of outstanding `read_lock`, and with a closed gate, and we encounter the originally presented issue.

^ I think the above is close to the situation we are seeing here.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1747,7 +1747,7 @@ void disk_log_impl::bg_checkpoint_offset_translator() {
     auto roll_lock_holder = co_await _segments_rolling_lock.get_units();
     auto t = term();
-    auto next_offset = offsets().dirty_offset + model::offset(1);
+    auto next_offset = model::next_offset(offsets().dirty_offset);
```
how is this related to the race?

WillemKauf replied:
Unrelated to the race, related to the test case I wrote in which I call `force_roll()` on a log that has nothing produced to it, `offsets.dirty_offset() == -9223372036854775808`, and the offset value `-9223372036854775808 + 1` is going to trigger an assert in [`disk_log_impl::new_segment()`](https://github.com/redpanda-data/redpanda/blob/60eadf0cb406ca16419537b96b819c8b6440e0e5/src/v/storage/disk_log_impl.cc#L1613-L1614).

If, instead, we use `model::next_offset(-9223372036854775808) == 0`, we are okay :+1:

dotnwat replied:
ahh, got it. i forgot that force_roll is test-only code. yolo!

WillemKauf commented:
Force push to:
* Change return statement on `_gate.closed()` to `_gate.check()` in `release_appender_in_background()`
* Wrap `log->force_roll()` in it's own function with a `try/catch` block in `storage_e2e_fixture_test.cc`

WillemKauf commented:
/ci-repeat 5
debug
skip-units
skip-redpanda-build
dt-repeat=100
tests/rptest/tests/availability_test.py

WillemKauf commented:
/ci-repeat 5
debug
skip-units
skip-redpanda-build
dt-repeat=100
tests/rptest/tests/availability_test.py

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x
